javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/RemappingUrlStreamHandler.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling;
+
+import static io.opentelemetry.javaagent.tooling.ShadingRemapper.rule;
+
+import io.opentelemetry.javaagent.bootstrap.InternalJarUrlConnection;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import net.bytebuddy.jar.asm.ClassReader;
+import net.bytebuddy.jar.asm.ClassWriter;
+import net.bytebuddy.jar.asm.commons.ClassRemapper;
+
+class RemappingUrlStreamHandler extends URLStreamHandler {
+
+  // We need to prefix the names to prevent the gradle shadowJar relocation rules from touching
+  // them. It's possible to do this by excluding this class from shading, but it may cause issue
+  // with transitive dependencies down the line.
+  private static final ShadingRemapper remapper =
+      new ShadingRemapper(
+          rule("#io.opentelemetry.api", "#io.opentelemetry.javaagent.shaded.io.opentelemetry.api"),
+          rule(
+              "#io.opentelemetry.context",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.context"),
+          rule(
+              "#io.opentelemetry.extension.aws",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.extension.aws"),
+          rule("#java.util.logging.Logger", "#io.opentelemetry.javaagent.bootstrap.PatchLogger"),
+          rule("#org.slf4j", "#io.opentelemetry.javaagent.slf4j"));
+
+  private final JarFile delegateJarFile;
+
+  public RemappingUrlStreamHandler(URL delegateJarFileLocation) {
+    try {
+      delegateJarFile = new JarFile(new File(delegateJarFileLocation.toURI()), false);
+    } catch (URISyntaxException | IOException e) {
+      throw new IllegalStateException("Unable to read internal jar", e);
+    }
+
+    System.out.println("Delegate jar file is " + delegateJarFile.getName());
+  }
+
+  /** {@inheritDoc} */
+  protected URLConnection openConnection(URL url) {
```

Comment:
are you sure that this is only called when class bytes are needed?

Reply:
Or resources. And we want to remap them as well, e.g. for Advices

Reply:
perhaps it would make sense to push reading bytes and remapping into InternalJarUrlConnection and do it only when bytes are really needed